packages/typespec-vscode/src/utils.ts

```diff
@@ -102,3 +93,211 @@ export async function loadModule(
     return undefined;
   }
 }
+
+export function tryParseJson(str: string): any | undefined {
+  try {
+    return JSON.parse(str);
+  } catch {
+    return undefined;
+  }
+}
+
+export async function tryReadFileOrUrl(
+  pathOrUrl: string,
+): Promise<{ content: string; url: string } | undefined> {
+  if (isUrl(pathOrUrl)) {
+    const result = await tryReadUrl(pathOrUrl);
+    return result;
+  } else {
+    const result = await tryReadFile(pathOrUrl);
+    return result ? { content: result, url: pathOrUrl } : undefined;
+  }
+}
+
+export async function tryReadFile(path: string): Promise<string | undefined> {
+  try {
+    const content = await readFile(path, "utf-8");
+    return content;
+  } catch (e) {
+    logger.debug(`Failed to read file: ${path}`, [e]);
+    return undefined;
+  }
+}
+
+export async function tryReadUrl(
+  url: string,
+): Promise<{ content: string; url: string } | undefined> {
+  try {
+    const response = await fetch(url, { redirect: "follow" });
+    const content = await response.text();
+    return { content, url: response.url };
+  } catch (e) {
+    logger.debug(`Failed to fetch from url: ${url}`, [e]);
+    return undefined;
+  }
+}
+
+export interface ExecOutput {
+  stdout: string;
+  stderr: string;
+  exitCode: number;
+  error: any;
+  spawnOptions: SpawnOptions;
+}
+export interface spawnExecutionEvents {
+  onStdioOut?: (data: string) => void;
+  onStdioError?: (error: string) => void;
+  onError?: (error: any, stdout: string, stderr: string) => void;
+  onExit?: (code: number | null, stdout: string, stderror: string) => void;
+}
+
+/**
+ * The promise will be rejected if the process exits with non-zero code or error occurs. Please make sure the rejection is handled property with try-catch
+ *
+ * @param exe
+ * @param args
+ * @param cwd
+ * @returns
+ */
+export function spawnExecutionAndLogToOutput(
+  exe: string,
+  args: string[],
+  cwd: string,
+): Promise<ExecOutput> {
+  return spawnExecution(exe, args, cwd, {
+    onStdioOut: (data) => {
+      logger.info(data.trim());
+    },
+    onStdioError: (error) => {
+      logger.error(error.trim());
+    },
+    onError: (error) => {
+      if (error?.code === "ENOENT") {
+        logger.error(`Cannot find ${exe} executable. Make sure it can be found in your path.`);
+      }
+    },
+  });
+}
+
+/**
+ * The promise will be rejected if the process exits with non-zero code or error occurs. Please make sure the rejection is handled property with try-catch
+ *
+ * @param exe
+ * @param args
+ * @param cwd
+ * @param on
+ * @returns
+ */
+export function spawnExecution(
+  exe: string,
+  args: string[],
+  cwd: string,
+  on?: spawnExecutionEvents,
+): Promise<ExecOutput> {
+  const shell = process.platform === "win32";
+  const cmd = shell && exe.includes(" ") ? `"${exe}"` : exe;
+  let stdout = "";
+  let stderr = "";
+
+  const options: SpawnOptions = {
```

Comment:
Can options: SpawnOptions be the function parameter since it is a common library, and we need to configure spawn options?
And remove `cwd` parameter

Reply:
any options you want to use? I can make the param an interface so that it's easier to extend in the future, but may not be suitable to expose the SpawnOptions directly because some option is actually not expected to use other value (i.e. shell).